semantics/cpp14/language/translation/decl/enum.k

```diff
@@ -0,0 +1,218 @@
+
+     syntax KItem ::= #addEnumToEnv(CId, Scope, CPPEnumType)
+     rule <k> #addEnumToEnv(X::CId, namespaceScope(N::Namespace), E::CPPEnumType) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          <ns-id> N </ns-id>
+          <ntypes> NT::Map => NT[X, Enum() <- E] </ntypes>
+
+     syntax KItem ::= enumContext(CId, CPPEnumType, List, fixed: Bool, ut: CPPIntegerType) [strict(2)]
+
+     rule enumContext(X::CId, E::CPPEnumType, Enumerators:List, Fixed::Bool, UT::CPPIntegerType)
+          => newEnum(E, Fixed)
+          ~> addEnumToEnv(X, E)
+          ~> scope(enumScope(getEnumId(E)), declareEnumerators(E, 0, 0, 0, int, Enumerators, Fixed, true))
+
+     // second param -underlying type
+     syntax KItem ::= newEnum(CPPEnumType, fixed: Bool)
+
+     rule <k> newEnum(E:CPPEnumType, true) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          (.Bag => <cppenum>...
+               <enum-id> getEnumId(E) </enum-id>
+               <enum-type> E </enum-type>
+               <scoped> getScoped(E) </scoped>
+               <enum-complete> true </enum-complete>
+               <enum-min> min(underlyingType(E)) </enum-min>
+               <enum-max> max(underlyingType(E)) </enum-max>
+           ...</cppenum>)
+
+     rule <k> newEnum(E:CPPUnscopedEnumType, false) => . ... </k>
+          <curr-tu> Tu::String </curr-tu>
+          <tu-id> Tu </tu-id>
+          (.Bag => <cppenum>...
+               <enum-id> getEnumId(E) </enum-id>
+               <enum-type> E </enum-type>
+               <scoped> false </scoped>
+               <enum-complete> false </enum-complete>
+           ...</cppenum>)
+
+     syntax KItem ::= declareEnumerators(CPPEnumType, min: Int, max: Int, next: Int, nextType: CPPSimpleType, List, fixed: Bool, first: Bool)
+
+     rule declareEnumerators(_, _, _, Next::Int, NextType::CPPSimpleType, ListItem(Enumerator(_, NoExpression() => prv(Next, noTrace, type(NextType)))) _::List, _, _)
+     context declareEnumerators(_, _, _, _, _, ListItem(Enumerator(_, HOLE:Expr)) _::List, _, _)
+
+     // Type of an enumerator inside the declaration is enum's underlying type
+     // if it is fixed, and type of initializing expression otherwise,
+     // Hence we need to store the type of underlying expression...
+     rule declareEnumerators(T::CPPEnumType, Min::Int, Max::Int, Next::Int, _, ListItem(Enumerator(X::CId, prv(V::Int, _, t(_, _, EType::CPPSimpleType)) #as Pr::PRVal)) L::List, Fixed::Bool, true)
+          => declareEnumerator(getEnumId(T), X, Pr, Fixed, underlyingType(T))
+          ~> declareEnumerators(T, V, V, V +Int 1, chooseEnumeratorType(EType, V +Int 1), L, Fixed, false)
+
+     rule declareEnumerators(T::CPPEnumType, Min::Int, Max::Int, Next::Int, _, ListItem(Enumerator(X::CId, prv(V::Int, _, t(_, _, EType::CPPSimpleType)) #as Pr::PRVal)) L::List, Fixed::Bool, false)
```

Comment:
I would prefer that instead of calling chooseEnumerationType we add a side condition that V +Int 1 is in the range of the type of the preceding enumerator, and if it isn't, then we generate an UNSPEC warning